File: solutions/haskell/acronym/5/src/Acronym.hs
module Acronym (abbreviate) where

import qualified Data.Text as T
import qualified Data.Char as Ch

abbreviate :: String -> String
abbreviate xs = (getOnlyUppers . splitInWords) textToAbreviate 
    where 
        textToAbreviate = T.pack xs
        splitInWords text = T.split (\ch -> ch `elem`  " -_") text
        getOnlyUppers text = T.unpack $ T.concat $ map getLetters text
        getLetters text
            | text == T.empty = text
            | T.all Ch.isUpper text = T.singleton $ Ch.toUpper $ T.head text 
            | otherwise = T.cons (Ch.toUpper $ T.head text) (T.filter Ch.isUpper (T.tail text))   
        

File: solutions/fsharp/acronym/2/Acronym.fs
module Acronym

let regex (pat: string) (str: string) =
    System.Text.RegularExpressions.Regex.IsMatch(str, pat)

let abbreviate (phrase: string) =

    phrase.Split [| ' '; '\t'; '-'; '\n' |]
    |> Seq.map (String.filter (string >> regex "[A-Za-z]"))
    |> Seq.filter (regex "^\s*$" >> not)
    |> Seq.map (Seq.head >> System.Char.ToUpper >> string)
    |> String.concat ""
